### ISSUE TITLE
coverage: identify contracts by hash, not bytecode

### ISSUE THREAD

**Issue by DaniPopes** [foundry-rs/foundry#9443]
### Component

Forge

### Describe the feature you would like

We already collect coverage hitmaps by bytecode hash, we can use this to avoid the hacky heuristics of finding a matching artifact using the bytecode itself.

In this loop the key is the bytecode hash, which is currently ignored: https://github.com/foundry-rs/foundry/blob/7a23a5cf851b991bfd2fde32d4f088319bbc1183/crates/forge/bin/cmd/coverage.rs#L253

cc @klkvr @grandizzy

**Comment by klkvr** [comment]
I don't think we can rely on pre-computed bytecode hashes for artifacts due to immutables?
